### PR TITLE
SECURITY-829 + fixing form usage

### PR DESCRIPTION
**Pull request by timja**
TODO:

* ~interface hudson.plugins.im.IMMessageTarget is missing its descriptor~
* ~Apply same changes to job configuration~
* ~WARNING	hudson.Functions#getPasswordValue: <f:password/> form control in Jenkins/configure.jelly IrcPublisher/DescriptorImpl/global.jelly is not backed by hudson.util.Secret~
* ~verify passwords are working I think encrypted value is saved in string, probably solved by above~

Requires:

https://github.com/jenkinsci/instant-messaging-plugin/pull/44
and
https://github.com/jenkinsci/instant-messaging-plugin/pull/43

should be fully compatible

Also looks like it provides jcasc support, it's in the export, haven't verified if it fully works

Note: I'm not a user of this plugin, I've tested form binding interactively in a debugger and all looks good but I haven't configured it to talk to an actual IRC server.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -1,139 +1,149 @@
...
-                <scope>import</scope>
-                <type>pom</type>
-            </dependency>
-        </dependencies>
-    </dependencyManagement>
+  <properties>
+    <revision>2.34</revision>
+    <changelist>-SNAPSHOT</changelist>
+    <gitHubRepo>jenkinsci/ircbot-plugin</gitHubRepo>
+    <version.instant-messaging.plugin>1.40-rc459.252399e42dd7</version.instant-messaging.plugin>
```
Comment (jimklimov) [inline pom.xml:29]: I suppose this references the suitable incrementals build? Should it be modified after "official release" of the updated dependency?

File: pom.xml [hunk 2]
```diff
@@ -1,139 +1,149 @@
...
-                <groupId>io.jenkins.tools.bom</groupId>
-                <artifactId>bom-2.204.x</artifactId>
-                <version>17</version>
-                <scope>import</scope>
-                <type>pom</type>
-            </dependency>
-        </dependencies>
-    </dependencyManagement>
+  <properties>
+    <revision>2.34</revision>
```
Comment (jimklimov) [inline pom.xml:26]: Probably a mis-merge from fixing over an older master commit? :) Probably related to the different white-spacing applied to the file here and in master.

File: pom.xml [hunk 3]
```diff
@@ -1,139 +1,149 @@
...
+    </repository>
+  </repositories>
 
-    <developers>
-        <developer>
-            <id>jimklimov</id>
-            <name>Jim Klimov</name>
-            <email>jimklimov+jenkinsci@gmail.com</email>
-            <timezone>1</timezone>
-        </developer>
```
Comment (jimklimov) [inline pom.xml:90]: Whyyyy? I was so young! (Also probably due to hard comparisons with the shifted whitespace)

**Comment by jimklimov** [comment]
I'll try to amend the pom.xml comments and whitespacing to match the master version.

By the way, before I get into testing - have you got any clue whether would users need to somehow edit and save configs to move their plaintext passwords into hashes? Or would it happen automatically by plugin upgrade alone? And would the plaintext copies of passwords disappear then so the original security issue is solved? Or would passwords just disappear and users must remember them (or find in backups) to reconfigure the IRC client? :)

File: pom.xml (hunk 1 above)
Comment (timja) [reply pom.xml:29]: yes

File: src/main/java/hudson/plugins/ircbot/IrcPublisher.java [hunk 4]
```diff
@@ -173,27 +171,24 @@ protected Object readResolve() {
 
         private boolean sslTrustAllCertificates;
 
+        @Deprecated
         String password = null;
```
Comment (timja) [inline src/main/java/hudson/plugins/ircbot/IrcPublisher.java:175]:
```suggestion
        transient String password = null;
```

File: src/main/java/hudson/plugins/ircbot/IrcPublisher.java [hunk 5]
```diff
@@ -173,27 +171,24 @@ protected Object readResolve() {
...
+        Secret secretPassword;
 
         private boolean sasl;
 
         String nick = "jenkins-bot";
 
+        @Deprecated
         String nickServPassword = null;
```
Comment (timja) [inline src/main/java/hudson/plugins/ircbot/IrcPublisher.java:183]:
```suggestion
        transient String nickServPassword = null;
```

**Comment by timja** [comment]
> I'll try to amend the pom.xml comments and whitespacing to match the master version.

I didn't intentionally change the whitespacing but yeah probably inconsistent spacing caused it.

> By the way, before I get into testing - have you got any clue whether would users need to somehow edit and save configs to move their plaintext passwords into hashes? Or would it happen automatically by plugin upgrade alone? And would the plaintext copies of passwords disappear then so the original security issue is solved? Or would passwords just disappear and users must remember them (or find in backups) to reconfigure the IRC client? :)

In the current fix a one time save of job configuration / system configuration would be required to update

> And would the plaintext copies of passwords disappear then so the original security issue is solved? Or would passwords just disappear and users must remember them (or find in backups) to reconfigure the IRC client? :)

No I've added migration logic so it all should work with no user intervention required

File: pom.xml (hunk 3 above)
Comment (jimklimov) [reply pom.xml]: Fixed

File: pom.xml (hunk 2 above)
Comment (jimklimov) [reply pom.xml]: Fixed

**Comment by jimklimov** [comment]
Thanks a lot for the wide-spread effort to fix this, I hope to give it a shot in practice some upcoming night.

**Comment by olblak** [comment]
\o/

**Comment by jimklimov** [comment]
Re-posting from IRC chat:

(10:21:54) jimklimov: so far I'm playing with the updated ircbot related plugins combined privately from your 3 PRs
(10:22:26) jimklimov: it seems that all was OK until I saved the global config :)
(10:23:41) jimklimov: this did remove "password" tags and added "secretPassword" (similar for nickserv), and added a channel secretPassword value
(10:24:15) jimklimov: funnily though, the old password was a base64 value that contained another base64 in braces
(10:24:43) jimklimov: the new secret one is that content {base64} in plaintext :)
(10:25:30) jimklimov: and it does not connect (disconnected some minutes after config save) :\
(10:26:37) jimklimov: curiously, I get notifications (so it got the server password right), but the client did not get into the channel - probably it tries that newly appeared secretPassword in the channel section where there was none originally
(10:27:34) jimklimov: oh, and that password value, base64 in braces, yields me some binary garbage so can't get the plantext password from it easily
(10:27:47) jimklimov: seems like plugin's insecurity was overestimated =D
(10:28:01) jimklimov: or gotta dig in backups for older snaps
(10:28:41) timja: jimklimov: i renamed the field to keep compatibility
(10:29:02) timja: i'll try connect to an irc server to see what the issue is
(10:31:17) jimklimov: strangely, in our logs I see that for some years there was no password at all (so probably just another team channel we use was so hidden)
(10:31:40) jimklimov: since a couple of years ago, the scrambledPassword setting became true, with password tags empty
(10:33:27) jimklimov: and since this summer there are those base64's wrapping another base64
(10:33:44) jimklimov: but apparently this was not a problem for half a year
(10:39:02) jimklimov: Message: java.lang.NullPointerException: Key for channel #BIOS cannot be null
(10:41:03) jimklimov: 2020-12-09 09:40:30.559+0000 [id=1093509]       SEVERE  h.i.i.InstallUncaughtExceptionHandler$DefaultUncaughtExceptionHandler#uncaughtException: A thread (IRC Bot/1093509) died unexpectedly due to an uncaught exception, this may leave your Jenkins in a bad way and is usually indicative of a bug in the code.
(10:41:03) jimklimov: java.lang.IllegalArgumentException: Nickserv password cannot be empty
(10:41:13) jimklimov: spectacular :)
(10:44:58) timja: want to post on the PR the issues so i don't lose them?
(10:45:18) timja: sounds like nickserv empty password breaks and passwords for channels aren't working right?
(10:45:33) timja: although i assume the general case is no password required for channel
(10:48:59) jimklimov: tried spaces (one and many) for the passwords - in nickserv it exploded like posted above, in server... not sure, in channel it is treated as null (probably whitespace is chomped away in Java code)

**Comment by jimklimov** [comment]
Also as reported in IRC, I had a problem saving the (legacy freestyle) job which has an IRC Notification post-build step, throwing a `net.sf.json.JSONException: JSONObject["notificationTargets"] is not a JSONArray.` in the Web-UI.

IIRC the context is that there is (was?) a list of channels (or direct user IDs) where Jenkins should post reports about this job's results; I think it was to override the Channels list from global config, for where the bot is present. It worked for us that the global list is default for notifications if nothing is set in the job's list, maybe even I was the one who coded it :) ... I wonder if this PR and/or posted instant-messaging-plugin recent PRs dropped the bit.

The original old (and still current) config did not specify any targets at job level:

````
    <hudson.plugins.ircbot.IrcPublisher plugin="ircbot@2.33">
      <targets class="empty-list"/>
      <strategy>ALL</strategy>
      <notifyOnBuildStart>true</notifyOnBuildStart>
      <notifySuspects>false</notifySuspects>
      <notifyCulprits>false</notifyCulprits>
      <notifyFixers>false</notifyFixers>
      <notifyUpstreamCommitters>false</notifyUpstreamCommitters>
      <buildToChatNotifier class="hudson.plugins.im.build_notify.PrintFailingTestsBuildToChatNotifier" plugin="instant-messaging@1.39-SNAPSHOT"/>
      <matrixMultiplier>ONLY_CONFIGURATIONS</matrixMultiplier>
      <extraMessage/>
      <customMessage/>
      <channels/>
    </hudson.plugins.ircbot.IrcPublisher>
````

For a fun data-point, at least in these job configs, the other items I was editing were saved; probably helps that IRC is last in list :) But although Jenkins seems to know the other fields I've changed relevant to other plugins, the on-disk config.xml files were not updated.

**Comment by jimklimov** [comment]
Also I think it was on IRC but not posted here, if there is a problem connecting to the configured IRC server, the code gets an NPE somewhere:

````
2020-12-09 21:30:35.652+0000 [id=21]    INFO    hudson.WebAppMain$3#run: Jenkins is fully up and running
2020-12-09 21:30:59.257+0000 [id=44]    INFO    h.p.i.IMConnectionProvider$ConnectorRunnable#run: Trying to reconnect
2020-12-09 21:30:59.374+0000 [id=44]    SEVERE  h.i.i.InstallUncaughtExceptionHandler$DefaultUncaughtExceptionHandler#uncaughtException: A thread (IM-Reconnector-Thread/44) died unexpectedly due to an uncaught exception, this may leave your Jenkins in a bad way and is usually indicative of a bug in the code.
java.lang.NullPointerException
        at hudson.plugins.ircbot.v2.IRCConnection.<init>(IRCConnection.java:105)
        at hudson.plugins.ircbot.v2.IRCConnectionProvider.createConnection(IRCConnectionProvider.java:36)
        at hudson.plugins.im.IMConnectionProvider.create(IMConnectionProvider.java:60)
        at hudson.plugins.im.IMConnectionProvider.access$500(IMConnectionProvider.java:17)
        at hudson.plugins.im.IMConnectionProvider$ConnectorRunnable.run(IMConnectionProvider.java:167)
        at java.lang.Thread.run(Thread.java:748)
````

This one is with a bad server IP/Port config. I think there was similar one for the case of accessible server, wrong password:

````
2020-12-09 09:40:30.556+0000 [id=112]   INFO    h.p.i.IMConnectionProvider$ConnectorRunnable#run:
  Trying to reconnect
2020-12-09 09:40:30.557+0000 [id=112]   INFO    h.p.ircbot.v2.IRCConnection#connect:
  Connecting to ircsrv:6667 as jenkins2 using charset UTF-8
2020-12-09 09:40:30.559+0000 [id=1093509]       SEVERE
  h.i.i.InstallUncaughtExceptionHandler$DefaultUncaughtExceptionHandler#uncaughtException:
  A thread (IRC Bot/1093509) died unexpectedly due to an uncaught exception, this may leave
  your Jenkins in a bad way and is usually indicative of a bug in the code.
java.lang.IllegalArgumentException: Nickserv password cannot be empty
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:88)
        at org.pircbotx.Configuration.<init>(Configuration.java:144)
        at org.pircbotx.Configuration$Builder.buildConfiguration(Configuration.java:589)
        at hudson.plugins.ircbot.v2.IRCConnection$2.run(IRCConnection.java:210)

2020-12-09 09:42:30.558+0000 [id=112]   WARNING h.p.ircbot.v2.IRCConnection#connect:
  Time out waiting for connecting to irc
2020-12-09 09:42:30.558+0000 [id=112]   WARNING h.p.im.IMConnectionProvider#create:
  Could not create IM connection
hudson.plugins.im.IMException: Connection failed
        at hudson.plugins.ircbot.v2.IRCConnectionProvider.createConnection(IRCConnectionProvider.java:42)
        at hudson.plugins.im.IMConnectionProvider.create(IMConnectionProvider.java:60)
        at hudson.plugins.im.IMConnectionProvider.access$500(IMConnectionProvider.java:17)
        at hudson.plugins.im.IMConnectionProvider$ConnectorRunnable.run(IMConnectionProvider.java:167)
        at java.lang.Thread.run(Thread.java:748)
2020-12-09 09:42:30.559+0000 [id=112]   INFO    h.p.i.IMConnectionProvider$ConnectorRunnable#run:
  Reconnect failed. Next connection attempt in 1 minutes
````

**Comment by timja** [comment]
Thanks will check, the databinding should be fixed now

**Comment by jimklimov** [comment]
The port number is not saved, stays 6667 whatever I type.

**Comment by jimklimov** [comment]
But it seems legacy job saving in UI is fixed, the on-disk XML changes:

````
-    <hudson.plugins.ircbot.IrcPublisher plugin="ircbot@2.33">
-      <targets class="empty-list"/>
+    <hudson.plugins.ircbot.IrcPublisher plugin="ircbot@2.36-SNAPSHOT">
+      <targets/>
````

**Comment by jimklimov** [comment]
Looks great! I was about to merge and rush to publish, but "Jenkins Username" (hudsonLogin in XML) still is not saved from global config. Two other logins, host, port, checkboxes, encoding, channel names, additional channel entries - these are saved (at least in global config :) ).

**Comment by jimklimov** [comment]
In jobs configured with notifications and no specified channels, the bot does send build results messages to all channel items listed in the global config.

If I specify a channel in job config, it is used *instead* of the global config, as was before (I think).

So both these use-cases seem okay.

**Comment by jimklimov** [comment]
I think I'll merge what we have so far, to record the achieved success (update the readme, etc.) before we botch it somehow :)

**Pull request merged**